Remove `actor->is_muted` and replace with a flag

dipinhora opened the pull request:
`actor->is_muted` is a `uint8_t` but it is always treated as a `boolean`.
It is also always only modified from a single scheduler thread (either to mute
or unmute the actor) but read from other scheduler threads.

This PR removes `actor->is_muted`, splits `actor->flags` into `actor->sync_flags`
and `actor->internal_flags` and adds a new sync flag called `FLAG_MUTED`.
The logic in `ponyint_is_muted`, `ponyint_mute_actor` and `ponyint_unmute_actor`
has been updated to use this new flag. Additionally, the sync flag read/modify
logic has been updated to use acquire/release semantics instead of relaxed
semantics so that it will work correctly on weakly ordered architectures (i.e. arm).

This preserves the existing behavior and makes sure all flags that are used from
multiple scheduler threads (overloaded, under pressure, muted, pending destroy)
are defined as sync flags and fully concurrency safe. The remaining flags
used from only a single scheduler thread are defined as internal flags don't use
atomics any longer which should result in a minor performance boost due to less
use of atomic operations when they aren't required.

SeanTAllen commented:
@dipinhora its accessed from multiple schedulers. We check it on every message send to see if the sender should be muted because the receiver is.

dipinhora commented:
@SeanTAllen yes, that's what i noted (see bolded and highlighted section of the quote). It is read from other schedulers but it is only written from a single scheduler just like all of the other `flags`.

> It is also always only modified from a single scheduler thread (either to mute
or unmute the actor) but __*read from other scheduler threads*__

Maybe i'm missing something though. If that's the case, please clarify.

SeanTAllen commented:
Reading the value incorrectly can lead to deadlock. I believe this change to be unsafe.

dipinhora commented:
Thanks for the clarification.

i believe what you're saying could happen due to how the `set_flag`, `unset_flag` and `has_flag` functions internally use `memory_order_relaxed` instead of `memory_order_acquire`/`memory_order_release`/`memory_order_acq_rel` for atomic synchronization that are currently used for `actor->is_muted`.

Is that your understanding also? (note: i'm trying to confirm that i understand your concern before i move forward with a fix)

SeanTAllen commented:
I'm away from my computer at the moment so I don't remember what the protection is on flags at the moment but I remember is muted was added due to flags not having proper protection, but I think after that we also started using the overloaded flag in a way that probably isn't safe right now. I hope that helps, if not, I'll be at the computer this evening and can give answers that aren't from memory.

dipinhora commented:
That is helpful and mostly confirms what i understood the concern to be. Thanks.

dipinhora commented:
@SeanTAllen i believe the latest commit (https://github.com/ponylang/ponyc/pull/4135/commits/1e096a670bedb2d1d869659f18d20e0dda697644) should address the concerns about concurrency and reads of stale values for the various `flags` that are accessed concurrently from multiple scheduler threads. Please let me know if you have any questions or concerns about it.

dipinhora commented on src/libponyrt/actor/actor.h:
```diff
@@ -91,6 +80,8 @@ gc_t* ponyint_actor_gc(pony_actor_t* actor);
 
 heap_t* ponyint_actor_heap(pony_actor_t* actor);
 
+bool ponyint_actor_unscheduled(pony_actor_t* actor);
```
note: this will be removed if the `FLAG_UNSCHEDULED` logic is removed as is being discussed in https://github.com/ponylang/ponyc/pull/4137

dipinhora commented:
@SeanTAllen This is ready to be reviewed again (i've updated the PR description to account for the newer commit changes). The last two commits are code cleanup and performance related and can be removed/split into their own PRs if desired.

SeanTAllen commented on src/libponyrt/actor/actor.c:
```diff
@@ -19,31 +19,148 @@
+
+static void set_sync_flag(pony_actor_t* actor, uint8_t flag)
+{
+  uint8_t flags = atomic_load_explicit(&actor->sync_flags, memory_order_acquire);
```
were all the atomics based on what we were doing for is_muted before?

SeanTAllen replied:
im looking at the atomics and they are different for sync vs what is_muted was. you appear to have weakened this for all the sync ones, I'm not sure about the flags.

SeanTAllen commented on src/libponyrt/actor/actor.c:
```diff
@@ -19,31 +19,148 @@
+    // 2. the sender isn't overloaded or under pressure
+    // AND
+    // 3. we are sending to another actor (as compared to sending to self)
+    if(triggers_muting(to) &&
+      !has_sync_flag_any(ctx->current, SYNC_FLAG_OVERLOADED |
+        SYNC_FLAG_UNDER_PRESSURE) &&
+      ctx->current != to)
+    {
+      ponyint_sched_mute(ctx, ctx->current, to);
+    }
+  }
```
this logic is rather different compared to the previous, did you pull a lot of additional logic into here?

SeanTAllen replied:
this looks very different. than what you are replacing. this looks like a mistake to me.

SeanTAllen commented on src/libponyrt/actor/actor.c:
```diff
@@ -19,31 +19,148 @@
+  }
+}
+
+static void maybe_mute(pony_ctx_t* ctx, pony_actor_t* to)
```
this appears to be the logic for ponyint_maybe mute?

SeanTAllen replied:
note that ponyint_maybe_mute and maybe_mute where two different functions and you appear to have turned them all into ponyint_maybe_mute calls while "renaming it" to `maybe_mute` which isn't correct.

dipinhora replied:
`maybe_mute` was only called from one place and i removed/replaced the code and renamed `ponyint_maybe_mute` to `maybe_mute` as part of https://github.com/ponylang/ponyc/pull/4135/commits/121f0311830d5436489018f17fe72b1e82b0ee7c (one of those last two commits are code cleanup and performance related that can be removed/split into their own PRs if desired that i mentioned in my earlier comment).

Seems like my two cleanup/performance related commits are causing confusion in reviewing this PR. i think i'll revert them and create separate PRs once this once has been settled.

SeanTAllen reviewed (changes requested):
see comments for concerns about incorrect change in logic regard `maybe_mute` and `ponyinit_maybe_mute` as well what appears to be a weakening of the barriers for is_muted when it changed to being for sync flags.

dipinhora commented on src/libponyrt/actor/actor.c:
```diff
@@ -19,31 +19,148 @@
+        SYNC_FLAG_UNDER_PRESSURE) &&
+      ctx->current != to)
+    {
+      ponyint_sched_mute(ctx, ctx->current, to);
```
`ponyint_mute_actor` got renamed to `mute_actor` as part of https://github.com/ponylang/ponyc/pull/4135/commits/121f0311830d5436489018f17fe72b1e82b0ee7c (one of those last two commits are code cleanup and performance related that can be removed/split into their own PRs if desired that i mentioned in my earlier comment).

Seems like my two cleanup/performance related commits are causing confusion in reviewing this PR. i think i'll revert them and create separate PRs once this once has been settled.

dipinhora commented:
@SeanTAllen sorry, i got overzealous and included two commits that did some code cleanup and performance optimizations that resulted in confusion. i have reverted both of them now will revisit them at a later point in other PRs.

If you take another look you'll see that the confusing bits around `ponyint_actor_mute`/`ponyint_maybe_mute`/`maybe_mute` have all been reverted.

The only remaining concern of yours is re: weakening of atomic guarantees from `actor->is_muted` to `actor->sync_flags` which i believe i addressed in my response. Please take a look at it and let me know if you have any further questions/concerns or if you'd prefer i change the code to use an atomic RMW compare/exchange instead.

SeanTAllen reviewed (approved):
Ah yes, I see what happened with the atomics now. Makes sense, I didn't spend enough time looking. This looks good to me, I'm leaving until sync tomorrow so someone else can perhaps review.

jemc commented:
I haven't had a chance to review the code in detail, but I trust Sean's review.

SeanTAllen commented:
@dipinhora this needs to be updated to reflect conflicts from the flag unscheduled removal.

dipinhora commented:
@SeanTAllen i've rebased onto latest `main` and resolve the conflicts...
